vowpalwabbit/predict_data.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+#include "ccb_label.h"
+
+/*
+ * class predict_data
+ * Description:
+ *   This data structure manages access and lifetime of the features data used in various VW reductions.
+ *
+ *   All contained data types MUST instantiate a `clear()` function that readies the object for reuse.
+ *
+ *   Calling create() on an existing element will clear and overwrite the existing data
+ *
+ * Member functions:
+ *   T& get<T>();
+ *   const T& get<T>() const;
+ *   void clear();
+ *
+ * Data structure usage:
+ *   Accessing an instance of existing data:
+ *     features_data fd;
+ *     auto& data = fd.get<data_type>();
+ */
+
+
+class predict_data
+{
+ private:
+  CCB::predict_data _ccb_predict_data;
```

Comment:
Could we drive generating this (and associated getters) through a template and inheritance?

In other words:

```
class predict_data : predict_container<CCB:predict_data> /* ... */
{
}
```

And predict container looks like:

```
template <typename T>
class predict_container 
{
private:
   T _data;

public:
   inline T& get() { return _data; };
   inline const T& get() const { return _data; };

protected:
   inline void clear() { _data.clear(); };
};
```

Not completely sure how to deal with clear, but possibly through a variadic template (ignore the completely incorrect syntax - I wanted to get the idea out without being bogged down looking at syntax; will update):

```
template <typename T, ...>
class predict_data_frame : predict_container<T>
{
private:
   predict_data_frame<NextT, ...> next;

public:
   inline void clear() 
   {
      // chain to protected method
      ((predict_container<T>*)this)->clear(); // is there cleaner syntax?
      next.clear();
   }
}

Reply:
I worked out how to make the template machinery work, but it is not the prettiest thing in the world.

https://gist.github.com/lokitoth/5aa5695b44a32b306d2e6341ee0f8053